src/metabase/query_processor/middleware/optimize_temporal_filters.clj

```diff
@@ -0,0 +1,205 @@
+  [[_ n unit] datetime-field-unit]
+  [:relative-datetime (if (= n :current) 0 n) (or unit datetime-field-unit)])
+
+(defmethod temporal-value-upper-bound :relative-datetime
+  [[_ n unit] datetime-field-unit]
+  [:relative-datetime (inc (if (= n :current) 0 n)) (or unit datetime-field-unit)])
+
+(defmulti ^:private optimize-filter
+  "Optimize a filter clause agat a bucketed `:datetime-field` clause and `:absolute-datetime` or `:relative-datetime`
+  value by converting to an unbucketed range."
+  {:arglists '([clause])}
+  mbql.u/dispatch-by-clause-name-or-class)
+
+(defmethod optimize-filter :=
+  [[_ field temporal-value]]
+  (let [[_ _ datetime-field-unit] (mbql.u/match-one field :datetime-field)]
+    (when (field-and-temporal-value-have-compatible-units? field temporal-value)
+      (let [field' (change-datetime-field-unit-to-default field)]
+        [:and
+         [:>= field' (temporal-value-lower-bound temporal-value datetime-field-unit)]
+         [:< field'  (temporal-value-upper-bound temporal-value datetime-field-unit)]]))))
+
+(defmethod optimize-filter :!=
+  [filter-clause]
+  (mbql.u/negate-filter-clause ((get-method optimize-filter :=) filter-clause)))
+
+(defn- optimize-comparison-filter
+  [optimize-temporal-value-fn [filter-type field temporal-value] new-filter-type]
+  [new-filter-type
+   (change-datetime-field-unit-to-default field)
+   (optimize-temporal-value-fn temporal-value (datetime-field-unit field))])
+
+(defmethod optimize-filter :<
+  [filter-clause]
+  (optimize-comparison-filter temporal-value-lower-bound filter-clause :<))
+
+(defmethod optimize-filter :<=
+  [filter-clause]
+  (optimize-comparison-filter temporal-value-upper-bound filter-clause :<))
+
+(defmethod optimize-filter :>
+  [filter-clause]
+  (optimize-comparison-filter temporal-value-upper-bound filter-clause :>=))
+
+(defmethod optimize-filter :>=
+  [filter-clause]
+  (optimize-comparison-filter temporal-value-lower-bound filter-clause :>=))
+
+(defmethod optimize-filter :between
+  [[_ field lower-bound upper-bound]]
+  (let [field' (change-datetime-field-unit-to-default field)]
+    [:and
+     [:>= field' (temporal-value-lower-bound lower-bound (datetime-field-unit field))]
+     [:<  field' (temporal-value-upper-bound upper-bound (datetime-field-unit field))]]))
+
+(defn- optimize-temporal-filters* [query]
+  (mbql.u/replace query
+    (_ :guard (partial mbql.u/is-clause? (set (keys (methods optimize-filter)))))
+    (or (when (can-optimize-filter? &match)
+          (u/prog1 (optimize-filter &match)
+            (if <>
+              (when-not (= &match <>)
+                (log/tracef "Optimized filter %s to %s" (pr-str &match) (pr-str <>)))
+              ;; if for some reason `optimize-filter` doesn't return an optimized filter clause, log and error and use
+              ;; the original. `can-optimize-filter?` shouldn't have said we could optimize this filter in the first
+              ;; place
```

Comment:
For things like this, do we have a special way of marking/capturing the info for our own diagnostic purposes?

Reply:
It will show up in the console, so hopefully someone will spot the error if it happens. We don't have any way to record the frequency of this happening, tho -- that's something that would be good to record